tests/unit/test_solving.py

```diff
@@ -0,0 +1,186 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Iterable, Union
+
+import attrs
+import pytest
+
+import qrules.particle
+import qrules.quantum_numbers
+import qrules.system_control
+import qrules.transition
+from qrules.conservation_rules import (
+    GraphElementRule,
+    c_parity_conservation,
+    parity_conservation,
+    spin_magnitude_conservation,
+    spin_validity,
+)
+from qrules.quantum_numbers import EdgeQuantumNumbers, NodeQuantumNumbers
+from qrules.solving import CSPSolver, EdgeSettings, NodeSettings, QNProblemSet
+from qrules.topology import MutableTransition
+
+if TYPE_CHECKING:
+    from qrules.argument_handling import Rule
+
+
+EdgeQuantumNumberTypes = Union[
+    type[EdgeQuantumNumbers.pid],
+    type[EdgeQuantumNumbers.mass],
+    type[EdgeQuantumNumbers.width],
+    type[EdgeQuantumNumbers.spin_magnitude],
+    type[EdgeQuantumNumbers.spin_projection],
+    type[EdgeQuantumNumbers.charge],
+    type[EdgeQuantumNumbers.isospin_magnitude],
+    type[EdgeQuantumNumbers.isospin_projection],
+    type[EdgeQuantumNumbers.strangeness],
+    type[EdgeQuantumNumbers.charmness],
+    type[EdgeQuantumNumbers.bottomness],
+    type[EdgeQuantumNumbers.topness],
+    type[EdgeQuantumNumbers.baryon_number],
+    type[EdgeQuantumNumbers.electron_lepton_number],
+    type[EdgeQuantumNumbers.muon_lepton_number],
+    type[EdgeQuantumNumbers.tau_lepton_number],
+    type[EdgeQuantumNumbers.parity],
+    type[EdgeQuantumNumbers.c_parity],
+    type[EdgeQuantumNumbers.g_parity],
+]
+
+NodeQuantumNumberTypes = Union[
+    type[NodeQuantumNumbers.l_magnitude],
+    type[NodeQuantumNumbers.l_projection],
+    type[NodeQuantumNumbers.s_magnitude],
+    type[NodeQuantumNumbers.s_projection],
+    type[NodeQuantumNumbers.parity_prefactor],
+]
+
+
+def test_solve(
+    all_particles: qrules.particle.ParticleCollection,
+    quantum_number_problem_set: QNProblemSet,
+) -> None:
+    solver = CSPSolver(all_particles)
+    result = solver.find_solutions(quantum_number_problem_set)
+    assert len(result.solutions) == 19
+
+
+def test_solve_with_filtered_quantum_number_problem_set(
+    all_particles: qrules.particle.ParticleCollection,
+    quantum_number_problem_set: QNProblemSet,
+) -> None:
+    solver = CSPSolver(all_particles)
```

Comment:
I figured that `all_particles` still contains all edge properties. These are still inserted into intermediate edges of the found solutions, which causes the large number of properties.

But even if you give the `CSPSolver` a list of `dict`s with on only the required properties (like you did with the domain filter), you still end up with a large number of solutions because of the `pid`s. You can visualize this by rendering the result with

```python
src = qrules.io.asdot(solutions)
graphviz.Source(src).render("file", format="svg")
```